### PR TITLE
Logging and VTuber exclusivity

### LINKED ISSUE #6: Logging
# Feature
Logging for changes made to messages and users of a Discord server. This should ease the moderation process of servers.

## Acceptance Criteria
- [x] Sushi Bot shall output posts into a designated Discord channel when changes are made to messages or users of the server.
- [x] Sushi Bot shall allow server owners to set up logging in their servers.

## Additional Context
The base logging feature will be ported from Sushi Stats.
A command allowing server owners to set up logging will be implemented.
The logging will have to be expanded to allow for setup in multiple servers.

### LINKED ISSUE #24: VTuber exclusivity
# Feature
Sushi Bot is a bot intended for VTubers and their communities and should therefore only be found there.

## Acceptance Criteria
- [x] Sushi Bot shall leave any server it has joined that doesn't belong to a VTuber.
  - [x] This check must be performed on start.
  - [x] This check may be manually triggered via a command.
  - [x] This check must be performed when the bot joins a server.

### PR DESCRIPTION
**Pull request by MirageAegis**
# What Does This PR Do?
This PR adds configurable moderation logs, VTuber exclusivity, and a few other QoL changes to Sushi Bot.

## Acceptance Criteria Met
- Sushi Bot has `/config logs` for server administrators to configure moderation logs.
- Sushi Bot outputs moderation logs in servers subscribed to logs.
- Sushi Bot leaves every server that doesn't belong to a VTuber.
  - It's checked on start.
  - It can be checked on demand by Sushi Hub administrators.
  - It's checked on new servers that Sushi Bot joins.

## Additional Changes
- Blacklisted users have their reason prepended with "Sushi Bot Blacklist: ".
- Sushi Bot bans all blacklisted users upon joining new eligible servers.
- Sushi Bot logs errors in the error log in Sushi Hub.

## Related Issues
Closes #6 
Closes #24

### CODE REVIEW

File: src/commands/configuration/config-subcommands/logs.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    // Command execution
+    async execute(ctx: ChatInputCommandInteraction): Promise<void> {
+        // Default values for parameters
+
+        const channel: TextChannel = ctx.options.getChannel("channel") ?? null;
+
+        await ctx.deferReply();
+
+        // Get the server documetn from the database
```
Comment (zahatikoff) [inline src/commands/configuration/config-subcommands/logs.ts]: Typo

File: package.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
     "author": "Mirage Aegis",
     "license": "MIT",
     "dependencies": {
-        "discord.js": "^14.11.0",
+        "discord.js": "^14.12.1",
```
Comment (MirageAegis) [inline package.json:18]: **IMPORTANT!!**
The discord.js dependency has been updated from v14.11.0 to v14.12.1
Make sure to run `npm ci` to make a clean install of the updated dependency

File: src/commands/configuration/config.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    help: new EmbedBuilder()
+        .setTitle("Config")
+        .setDescription(
+            `Commands for configuring events in a server. Use \`/${name} help\` for information about these commands. Can only be used by server administrators`
+        )
+        .addFields(
+            { name: "Format", value: `\`/${name} <subcommand>\`` },
+            { name: "<subcommand>", value: "The subcommand to use" }
+        )
+};
```
Comment (MirageAegis) [inline src/commands/configuration/config.ts:91]: This is the top level `/config` command used for server configurations

File: src/commands/configuration/config-subcommands/help.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            return;
+        }
+    },
+
+    // Error handler
+    error: defaultErrorHandler,
+
+    // Help command embed
+    help: help
+};
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/help.ts:138]: This is the `/config help` which works similarly to `/help` and `/admin help`

File: src/commands/configuration/config-subcommands/logs.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    // Help command embed
+    help: new EmbedBuilder()
+        .setTitle("Logs")
+        .setDescription(
+            "A server configutation command that sets the logging channel for a server")
+        .addFields(
+            { name: "Format", value: `\`/config ${name} [logs]\`` },
+            { name: "[logs]", value: "Optional parameter. The channel to set as logging channel. Omit to remove the logs channel" }
+        )
+};
```
Comment (MirageAegis) [inline src/commands/configuration/config-subcommands/logs.ts]: This is the `/config logs` command which is used to set up moderation logs
It saves the logs channel's channel ID to the database

File: src/events/events.ts [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+import { Server } from "../schemas/server";
+import {
+    genMemberBanEmbed, genMemberJoinEmbed, genMemberLeaveEmbed, genMemberUnbanEmbed,
+    genMemberUpdateEmbed, genMessageDeleteEmbed, genMessageEditEmbed, genUserUpdateEmbed
+} from "./logs";
+import { Blacklist } from "../schemas/blacklist";
+
+/*
+ * This module has event listeners for the bot
+ */
```
Comment (MirageAegis) [inline src/events/events.ts:40]: This module has the event listeners used for moderation logs and VTuber exclusivity

File: src/events/events.ts [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+        // Log to the logs channel
+        await logs.send({ embeds: [genMemberUnbanEmbed(ban)] });
+    } catch {
+        // Remove configurations if logging failed.
+        // Most likely causes are that the channel was deleted or that
+        // Sushi Bot had its permissions revoked
+        server.logs = null;
+        await server.save();
+    }
+};
```
Comment (MirageAegis) [inline src/events/events.ts:171]: These four listeners are quite similar. They get triggered when a server gains or loses a member and logs it to the moderation logs of servers subscribed to logs

File: src/events/logs.ts [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+        })
+        .setThumbnail(ban.user.displayAvatarURL())
+        .addFields(
+            { name: "ID", value: ban.user.id, inline: true },
+            { name: "Created", value: `<t:${Math.floor(ban.user.createdTimestamp / millisToSecs)}>`, inline: true },
+            { name: "Bot user", value: `${ban.user.bot}`, inline: true }
+        );
+
+    return embed;
+};
```
Comment (MirageAegis) [inline src/events/logs.ts:131]: These are the embeds used by the listeners above

File: src/events/events.ts [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+        // Log to the logs channel
+        await logs.send({ embeds: [embed] });
+    } catch {
+        // Remove configurations if logging failed.
+        // Most likely causes are that the channel was deleted or that
+        // Sushi Bot had its permissions revoked
+        server.logs = null;
+        await server.save();
+    }
+};
```
Comment (MirageAegis) [inline src/events/events.ts:208]: This listener gets triggered when a server member updates their server profile

**Review by zahatikoff — DISMISSED** [review]

File: src/events/logs.ts [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+    // Add this if the avatar was updated
+    if (oldPfp !== newPfp) {
+        embed.addFields(
+            { name: "New avatar", value: "_ _" } // Empty field value
+        )
+            .setImage(after.displayAvatarURL());
+    }
+
+    return embed;
+};
```
Comment (MirageAegis) [inline src/events/logs.ts]: This is the embed used in the listener above
It checks for nickname and server avatar updates

File: src/commands/admin/admin-subcommands/reserver.ts
Comment (zahatikoff) [inline src/commands/admin/admin-subcommands/reserver.ts:1]: **Reserver** does not seem to be the most fitting name for the command, as it mostly just reloads the bot?

File: src/events/events.ts [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+            ]});
+        } catch {
+            // Remove configurations if logging failed.
+            // Most likely causes are that the channel was deleted or that
+            // Sushi Bot had its permissions revoked
+            server.logs = null;
+            await server.save();
+        }
+    }
+};
```
Comment (MirageAegis) [inline src/events/events.ts:265]: This listener gets triggered when a user makes changes to their global profile; and broadcasts the changes, if they are significant to Sushi Bot's logs, to all servers' logs channels

File: src/events/logs.ts [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+    // Add this if the username was updated
+    if (oldPfp !== newPfp) {
+        embed.addFields(
+            { name: "New avatar", value: "_ _", inline: false }
+        )
+            .setImage(after.displayAvatarURL());
+    }
+
+    return embed;
+};
```
Comment (MirageAegis) [inline src/events/logs.ts:244]: This is the embed used for the listener above
It checks for changes to usernames, display names, and avatars

File: src/events/events.ts [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+        const bl: Blacklist = await Blacklist.get();
+
+        for await (const [uid, reason] of bl.users) {
+            await server.bans.create(
+                uid,
+                { reason: reason }
+            );
+        }
+    }
+};
```
Comment (MirageAegis) [inline src/events/events.ts:361]: These two get triggered when the state of a message changes

File: src/events/logs.ts [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+            urls += `${attachment.url}\n`;
+        }
+
+        embed.addFields(
+            { name: "Attachments", value: urls, inline: false }
+        );
+    }
+
+    return embed;
+};
```
Comment (MirageAegis) [inline src/events/logs.ts:382]: These are the embeds used for message changes
They check the message's content and attachments

File: src/schemas/server.ts [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+ */
+
+import mongoose, { Schema, Document, InferSchemaType } from "mongoose";
+
+const serverSchema: Schema = new Schema({
+    _id: String,
+    logs: String,
+});
+
+type ServerT = InferSchemaType<typeof serverSchema>;
```
Comment (MirageAegis) [inline src/schemas/server.ts:32]: This is the database schema used to represent server configurations, more fields will be added in the future

File: src/schemas/server.ts [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+     * The ID of the logs channel
+     */
+    public get logs(): string {
+        return (<ServerT> this.data).logs;
+    }
+
+    public set logs(id: string) {
+        (<ServerT> this.data).logs = id;
+    }
+}
```
Comment (MirageAegis) [inline src/schemas/server.ts]: This is the wrapper class for the database schema, it has all the necessary functions and methods

File: src/schemas/server.ts [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+     * The corresponding Mongo model used for reading and writing to the database.
+     */
+    private static readonly model = mongoose.model("Server", serverSchema);
+
+    /**
+     * The cached servers.
+     * The servers are stored along with the ID for the timer responsible for
+     * clearing the memory.
+     */
+    private static cache: Map<string, [Server, NodeJS.Timeout]> = new Map();
```
Comment (MirageAegis) [inline src/schemas/server.ts:51]: The server configurations that are fetched from the database are cached to improve performance at the cost of memory

File: src/schemas/server.ts [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        const timeout: NodeJS.Timeout = setTimeout(async (): Promise<void> => {
+            await server.save();
+            Server.cache.delete(id);
+        }, serverLifetime);
+
+        // Save to the cache
+        Server.cache.set(id, [server, timeout]);
+
+        return server;
+    }
```
Comment (MirageAegis) [inline src/schemas/server.ts:125]: They are cached and refreshed in the `Server.get` function

File: src/events/events.ts [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+ * Triggered when the bot leaves a Discord server,
+ * or when a Discord server is deleted.
+ * It deletes the server configurations from the database.
+ * 
+ * @param client the Discord bot
+ * @param server the Discord server left
+ */
+export const onServerLeave = async (client: Client, server: Guild): Promise<void> => {
+    await (await Server.get(server.id)).delete();
+};
```
Comment (MirageAegis) [inline src/events/events.ts:373]: This gets triggered when Sushi Bot leaves a server
It deletes the server configurations for that server

File: src/events/events.ts [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+export const onError = async (client: Client, error: Error): Promise<void> => {
+    const logs: TextChannel = getAdminLogsChannel();
+
+    const report: string = "```\n" +
+                           "Generic Error\n\n" +
+                           `${error}\n` +
+                           "```";
+    
+    await logs.send(report);
+};
```
Comment (MirageAegis) [inline src/events/events.ts:391]: This is the error handler for generic errors
It logs the errors to the error log in the admin server

File: src/util/error-handler.ts [hunk 21]
```diff
@@ -74,4 +80,23 @@ export const defaultErrorHandler: ErrorHandler = async (ctx: ChatInputCommandInt
...
+                           "Command Error\n\n" +
+                           `Server: ${ctx.guild.name}\n` +
+                           `User: ${user.id} (${user.username})\n` +
+                           `Command: ${command.name}\n` +
+                           `Options: ${command.options.toString()}\n\n` +
+                           // Error name, and error code if it's a Discord API error
+                           `${err.name}${err instanceof DiscordAPIError ? `: ${err.code} (${err.message})` : ""}\n` +
+                           "```";
+    
+    await getAdminLogsChannel().send(report);
```
Comment (MirageAegis) [inline src/util/error-handler.ts:101]: This has been added to the command error handler to make it log to the error log in the admin server

File: src/util/init.ts [hunk 22]
```diff
@@ -54,8 +72,54 @@ export const init = async (client: Client): Promise<void> => {
...
+    });
+
+    client.on(Events.MessageDelete, async (message: Message): Promise<void> => {
+        await onMessageDelete(client, message);
+    });
+
+    client.on(Events.Error, async (error: Error): Promise<void> => {
+        await onError(client, error);
+    });
+};
```
Comment (MirageAegis) [inline src/util/init.ts:125]: This is where all the listeners are loaded into the bot

**Review by MirageAegis — COMMENTED** [review]
This review will walk through key points of this PR, which mainly consists of implementations for #6 and #24

File: src/commands/configuration/config-subcommands/logs.ts (hunk 1 above)
Comment (MirageAegis) [reply src/commands/configuration/config-subcommands/logs.ts]: I'll fix that

**Review by zahatikoff — APPROVED** [review]
I don't think the `/admin reserver` command name is fitting, but since it's internal use only it's fine for now

**Review by zahatikoff — DISMISSED** [review]

File: src/commands/admin/admin-subcommands/reserver.ts
Comment (MirageAegis) [reply src/commands/admin/admin-subcommands/reserver.ts:1]: It only reloads the servers that the bot is in but I can see why it doesn't seem very fitting. We can change it in the future if we come up with a better name, or we can merge the command with `/admin reblacklist` and call it `/admin reload`

Comment (zahatikoff) [reply src/commands/admin/admin-subcommands/reserver.ts:1]: Yeah also sorry for a possible shitload of notiffs
I got a lil confused with the code review tool

Comment (MirageAegis) [reply src/commands/admin/admin-subcommands/reserver.ts:1]: No worries!
I'll get on with the `/admin reload` command when I have time. In the meantime, you can go through the comments I made and resolve them as you read them

**Pull request merged**